Migrate to mono 5

marksvc opened the pull request:
* Add to instructions.
* Get tests working with mono-sil (mono 3).
* Migrate codebase to mono5-sil.

jasonleenaylor commented on src/ECInterfaces/Util.cs:
```diff
@@ -55,19 +55,20 @@ private static string CommonApplicationData
+                            s_CommonAppDataFolder = "/var/lib/encConverters";
+                        #endif
+                        DebugWriteLine(typeof(Util).ToString(), "CommonAppDataFolder = " + s_CommonAppDataFolder);
+                    } else {
```
else should be on new line

marksvc replied:
Done

jasonleenaylor commented on src/EncCnvtrs/EncCnvtrs 2010.csproj:
```diff
@@ -146,7 +146,7 @@
     <Reference Include="ECInterfaces, Version=4.0.0.0, Culture=neutral, PublicKeyToken=f1447bae1e63f485, processorArchitecture=x86">
       <SpecificVersion>False</SpecificVersion>
-      <HintPath>..\..\output\release\ECInterfaces.dll</HintPath>
+      <HintPath>..\..\output\$(Configuration)\ECInterfaces.dll</HintPath>
```
This shouldn't be necessary, output paths for both release and debug will be checked automatically by msbuild. Also I'm not sure this is valid for the HintPath in VisualStudio (I'd have to check and see).

marksvc replied:
IIRC, the reason for this was because VSCode was using referenced libraries to resolve symbols and determine intellisense. And this allowed it to do that for some code I was working with. 
Notice, that this isn't even the actual project file that mono would be using (which would presumably be `EncCnvtrs-mono.csproj`). 
I did not look into why VSCode was using this `EncCnvtrs 2010.csproj` instead of the `-mono.csproj` file. I assume it was looking at a `.sln` file, or maybe picking the first `.csproj` in the directory. 
Now that I have edited the code in question, I won't be benefiting from VSCode's intellisense etc unless I revisit that area, and so I could change this back if it's thought to be a problem.

jasonleenaylor commented on src/IcuEC/IcuTranslitAutoConfigDialog.cs:
```diff
@@ -22,23 +22,45 @@ public class IcuTranslitAutoConfigDialog : SilEncConverters40.AutoConfigDialog
+        }
+
+        [DllImport("IcuTranslitEC.dll", EntryPoint = "IcuTranslitEC_ConverterNameList_next", CallingConvention = CallingConvention.Cdecl)]
+		[return : MarshalAs(UnmanagedType.LPStr)]
```
Some indent problems here, check the whole file.

marksvc replied:
Done

ermshiperete commented on ReadMe.md:
```diff
@@ -26,5 +42,4 @@ In Windows,
 2. Increase the version in Package.nuspec
-3. Run 'NuGet.exe pack' from the root directory of the repository
-
+3. Run 'NuGet.exe pack' from the root directory of the repository
```
That's `nuget.exe` nowadays...

marksvc replied:
Thanks. Done.

ermshiperete commented on build/EC.proj:
```diff
@@ -59,63 +59,63 @@
 				 Targets="$(msbuild-target)"
 				 Properties="$(msbuild-props)"
-				 ToolsVersion="3.5"/>
+				 ToolsVersion="15.0"/>
```
We should omit `ToolsVersion`. Specifying it causes problems with Mono 6, and according to Microsoft developers it's not needed and doesn't do what you think it does. So best to leave it out.

marksvc replied:
Done

ermshiperete commented on src/IcuEC/IcuConvAutoConfigDialog.cs:
```diff
@@ -19,24 +19,47 @@ public class IcuConvAutoConfigDialog : SilEncConverters40.AutoConfigDialog
 
-#if __MonoCS__
-		[DllImport("IcuConvEC.dll", EntryPoint = "IcuConvEC_ConverterNameList_next", CallingConvention = CallingConvention.Cdecl)]
+        static string CppConverterNameList_next()
```
Wrong indentation - but I guess Jason already mentioned that.

marksvc replied:
Done

ermshiperete commented on src/IcuEC/IcuConvEncConverter.cs:
```diff
@@ -33,24 +33,45 @@ public class IcuConvEncConverter : EncConverter
 
-#if __MonoCS__
-		[DllImport("IcuConvEC.dll", EntryPoint = "IcuConvEC_ConverterNameList_next", CallingConvention = CallingConvention.Cdecl)]
+        static string CppConverterNameList_next()
```
Indentation

marksvc replied:
Done

ermshiperete commented on src/IcuEC/IcuConvEncConverter.cs:
```diff
@@ -161,30 +182,36 @@ protected unsafe void Load(string strConvID)
-				// legacy
-				eInFormEngine = EncodingForm.LegacyBytes;
-			}
+            {
```
Indentation

marksvc replied:
Done

ermshiperete commented on src/IcuEC/IcuTranslitEncConverter.cs:
```diff
@@ -37,24 +37,45 @@ public class IcuTranslitEncConverter : EncConverter
-		[DllImport("IcuTranslitEC.dll", EntryPoint="IcuTranslitEC_ConverterNameList_next", CallingConvention = CallingConvention.Cdecl)]
-		[return: MarshalAs(UnmanagedType.LPStr)]
-		static extern string CppConverterNameList_next();
+        static string CppConverterNameList_next()
```
Indentation

marksvc replied:
Done

ermshiperete commented on src/PyScriptEC/PyScriptEncConverter.cs:
```diff
@@ -129,13 +129,16 @@ bool                bForward
-                Util.DebugWriteLine(this, "eInFormEngine UTF32");
-                eInFormEngine = EncodingForm.UTF32;
-#endif
+                if (!Util.IsUnix) // _MSC_VER
```
It would be easier to if you'd switch the two cases and put `Util.IsUnix` first.

marksvc replied:
Why is it easier? Done.

ermshiperete replied:
It's easy to overlook the `!`, and the two blocks don't make it obvious which one is for which platform. 
Usually I think it's recommended practice if you have both `if` and `else` that you order the blocks in such a way that you don't need the negation.

marksvc replied:
That makes sense, thanks.

ermshiperete commented on src/PyScriptEC/PyScriptEncConverter.cs:
```diff
@@ -129,13 +129,16 @@ bool                bForward
                 // We could use UTF-8 here, but wide data works just fine.
                 // the windows version definitely needs UTF16, but for some reason _MSC_VER is not defined on windows (as I think Jim is expecting)
                 //  so I'll use the opposite logic of 'if not mono'...
```
The comment is slightly outdated now. Could be something like

```
// We could use UTF-8 here, but wide data works just fine. The windows version definitely needs UTF16.
```

marksvc replied:
Right, thank you.

marksvc replied:
Done

ermshiperete commented on src/PyScriptEC/PyScriptEncConverter.cs:
```diff
@@ -146,13 +149,16 @@ bool                bForward
-                Util.DebugWriteLine(this, "eOutFormEngine UTF32");
-                eOutFormEngine = EncodingForm.UTF32;
-#endif
+                if (!Util.IsUnix) // _MSC_VER
```
see above

marksvc replied:
Done.

marksvc commented:
I made changes based on code review comments. I may not have adequately resolved a comment about $(Configuration) in HintPath. 
I added commits and also changed one or two.
23/23 unit tests pass on my workstation.
If possible, I would be glad for this PR to be merged or rebased but not squashed, so that individual pieces of work can be examined later more easily.
